Diabetes Prediction Using Support Vector Machine(Supervised Machine Learning Model) : **Mohammad Wasiq**

Importing the Dependencies:

In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score
import pickle

Data Collection and Analysis:

PIMA Diabetes Dataset

In [2]:
#loading the data into pandas dataframe

diabetes_df = pd.read_csv('diabetes.csv')

In [3]:
#printing the first five rows of the dataset

diabetes_df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [4]:
#this function give the information regarding any particular function what parameters should be passed and all


pd.read_csv?



In [5]:
# getting the number of rows and colums in the dataset

diabetes_df.shape

(768, 9)

In [6]:
# getting the statistical measures of the data

diabetes_df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [7]:
diabetes_df['Outcome'].value_counts()

0    500
1    268
Name: Outcome, dtype: int64

0 --> Non Diabetic Patients.


1 --> Diabetic Patients.

Here we can see that there are 500 people who are non diabetic and 268 people who are diabetic. 

In [8]:
# getting the mean value of each feature for the Outcome lebel

diabetes_df.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

Here we can observe that people who are diabetic have very high glucose level as well as Insulin level than the ones who are non diabetic.

Seperating the data(features) and the lebel(target):

In [9]:
X = diabetes_df.drop(columns= 'Outcome', axis= 1)
Y = diabetes_df['Outcome']

In [10]:
print(X)

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  


In [11]:
print(Y)

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


DATA STANDARISATION:

In [12]:
scaler = StandardScaler()

In [13]:
standarized_data = scaler.fit_transform(X)

In [14]:
print(standarized_data)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [15]:
X = standarized_data
Y = diabetes_df['Outcome']

In [16]:
print(X)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [17]:
print(Y)

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


Splitting the data into training data and test data using Train Test Split Function:

In [18]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=1)

In [19]:
print(X_train.shape)

(614, 8)


In [20]:
print(X_test.shape)

(154, 8)


In [21]:
print(Y_train.shape)

(614,)


In [22]:
print(Y_train.shape)

(614,)


In [23]:
print(X.shape, X_train.shape, X_test.shape)

(768, 8) (614, 8) (154, 8)


Training the model:

In [24]:
classifier = svm.SVC(kernel = 'linear')

In [25]:
#training the support vector machine Classifier

diabetesmodel = classifier.fit(X_train, Y_train)



Model Evaluation by calculating the accuracy of the trained model:

In [26]:
#accuracy score on the training data

X_train_prediction = classifier.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [27]:
print('Accuracy score of the training data: ' ,training_data_accuracy)

Accuracy score of the training data:  0.7833876221498371


In [28]:
#accuracy score on the test data

X_test_prediction = classifier.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)
print('Accuracy score of the test data: ' ,test_data_accuracy)

Accuracy score of the test data:  0.7792207792207793


Making the Predictive System for our model to start predicting:

In [29]:
input_data = (15,136,70,32,110,37.1,0.153,43)

#coverting the input data to a numpy array

input_data_as_numpy_array = np.asarray(input_data)

# reashaping the array as we are predicting for just one instance

input_data_reshaped = input_data_as_numpy_array.reshape(1, -1)

#standarize the input data

std_data = scaler.transform(input_data_reshaped)


prediction = classifier.predict(std_data)


if (prediction[0] == 0):
  print('The person is non diabetic.')
else:
  print('The person is diabetic.')

The person is diabetic.


/Users/tarunsingh/opt/anaconda3/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [30]:
#open a file where you want to store the data

file = open ('diabetesmodel.pkl', 'wb')

#dump information to that file
pickle.dump(diabetesmodel, file)